modules/storages/app/common/storages/peripherals/storage_interaction/authentication.rb

```diff
@@ -0,0 +1,140 @@
+# frozen_string_literal:true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2024 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Storages
+  module Peripherals
+    module StorageInteraction
+      module Authentication
+        class << self
+          def with_basic_auth(storage:, http_options: {})
+            username = storage.username
+            password = storage.password
+
+            if username.blank? || password.blank?
+              log_message = 'Cannot authenticate storage with basic auth. Password or username not configured.'
+              return error(code: :error, log_message:, payload: storage)
+            end
+
+            yield OpenProject.httpx.basic_auth(username, password).with(http_options)
+          end
+
+          def with_user_token(storage:, user:, http_options: {}, &)
+            config = storage.oauth_configuration
+
+            current_token = OAuthClientToken.find_by(user_id: user, oauth_client_id: config.oauth_client.id)
+            if current_token.nil?
+              return error(code: :unauthorized,
```

Comment:
This can be misleading. It isn't unauthorized, it was even tried. :P

Saying that the user isn't authenticated against storage X would be more precise.

Reply:
actually, that is the only one, that is taken over from previous state and me thinks, it makes sense.

if you never did the login, your are `unauthorized`. It is similar to e.g. how HTTP reacts to the absence of a `Authorization` header.